apps/arweave/src/ar_verify_chunks.erl

```diff
@@ -299,6 +321,59 @@ report_progress(State) ->
 	ar_verify_chunks_reporter:update(StoreID, Report2),
 	State#state{ verify_report = Report2 }.
 
+%% Generate a list of Count random offsets in the range (Start, End]
+%% (i.e. offsets strictly greater than Start and less than or equal to End)
+%% such that if an offset is sampled, no other offsets are selected from the
+%% open interval (Offset - ?DATA_CHUNK_SIZE, Offset + ?DATA_CHUNK_SIZE).
+generate_sample_offsets(Start, End, Count) when is_integer(Start), is_integer(End) ->
+	Candidates = lists:seq(Start + 1, End, ?DATA_CHUNK_SIZE),
+	Count2 = min(Count, length(Candidates)),
+	pick_offsets(Candidates, Count2).
+
+pick_offsets(_Candidates, 0) ->
+	[];
+pick_offsets([], _Count) ->
+	[];
+pick_offsets(Candidates, Count) ->
+	Offsets = pick_offsets_batch(min(Count, length(Candidates)), Candidates),
+	UniqueOffsets = lists:usort(Offsets),
+	NewCandidates = Candidates -- UniqueOffsets,
+	UniqueOffsets ++ pick_offsets(NewCandidates, Count - length(UniqueOffsets)).
+
+pick_offsets_batch(Len, Candidates) ->
+	pick_offsets_batch(Len, Candidates, min(Len, 20)).
+
+pick_offsets_batch(_Len, _Candidates, 0) ->
+	[];
+pick_offsets_batch(Len, Candidates, BatchSize) ->
+	N = rand:uniform(Len),
+	[lists:nth(N, Candidates) | pick_offsets_batch(Len, Candidates, BatchSize - 1)].
+
+%% Use generate_sample_offsets/3 to obtain offsets (with exclusion)
+%% and then queries ar_data_sync:get_chunk/2 with options to trigger unpacking.
+sample_random_chunks(Count, Packing, Start, End, StoreID) ->
+	Offsets = generate_sample_offsets(Start, End, Count),
```

Comment:
For a full 3.6TB partition I think Offsets will be a list of length 14.4 million entries or have I misread?